File: Iteracion 2 version final/iteracion 2/loader.c
#include<stdio.h> 
#include <stdlib.h>
#include <string.h>

#include "parser.c"

//Machine
#include "Machine.c"



int getOperationCode( const char * operation ){
   
   int opCode = 0;

    if( strcmp(operation,"LOAD") == 0)
        opCode = LOADop;
    else if( strcmp(operation,"LOADA") == 0)
        opCode = LOADAop;
    else if( strcmp(operation,"LOADI") == 0 )
        opCode = LOADIop;
    else if( strcmp(operation,"LOADL") == 0 )
        opCode = LOADLop;
    else if( strcmp(operation,"STORE") == 0 )
        opCode = STOREop;
    else if( strcmp(operation,"STOREI") == 0 )
        opCode = STOREIop;
    else if( strcmp(operation,"CALL") == 0 )
        opCode = CALLop;
    else if( strcmp(operation,"CALLI") == 0 )
        opCode = CALLIop;
    else if( strcmp(operation,"RETURN") == 0 )
        opCode = RETURNop;
    else if( strcmp(operation,"PUSH") == 0 )
        opCode = PUSHop;
    else if( strcmp(operation,"POP") == 0 )
        opCode = POPop;
    else if( strcmp(operation,"JUMP") == 0 )
        opCode = JUMPop;
    else if( strcmp(operation,"JUMPI") == 0 )
        opCode = JUMPIop;
    else if( strcmp(operation,"JUMPIF") == 0 )
        opCode = JUMPIFop;
    else if( strcmp(operation,"HALT") == 0 )
        opCode = HALTop;

   return opCode;
   }




int getRegisterCode( const char * registers ){
   

   int registerCode;
   
    if( strcmp(registers,"CB") == 0)
        registerCode = CBr;
    else if( strcmp(registers,"CT") == 0)
        registerCode = CTr;
    else if( strcmp(registers,"PB") == 0)
        registerCode = PBr;
    else if( strcmp(registers,"PT") == 0)
        registerCode = PTr;
    else if( strcmp(registers,"SB") == 0)
        registerCode = SBr;
    else if( strcmp(registers,"ST") == 0)
        registerCode = STr;
    else if( strcmp(registers,"HB") == 0)
        registerCode = HBr;
    else if( strcmp(registers,"HT") == 0)
        registerCode = HTr;
    else if( strcmp(registers,"LB") == 0)
        registerCode = LBr;
    else if( strcmp(registers,"L1") == 0)
        registerCode = L1r;
    else if( strcmp(registers,"L2") == 0)
        registerCode = L2r;
    else if( strcmp(registers,"L3") == 0)
        registerCode = L3r;
    else if( strcmp(registers,"L4") == 0)
        registerCode = L4r;
    else if( strcmp(registers,"L5") == 0)
        registerCode = L5r;
    else if( strcmp(registers,"L6") == 0)
        registerCode = L6r;
    else if( strcmp(registers,"CP") == 0)
        registerCode = CPr; 
    else
        registerCode = -1;
   
   return registerCode;
}

   void writeR (char leftbracket, int r, char rightbracket) {

    printf("%c",leftbracket);
    switch (r) {
      case CBr:
        printf("CB");
        break;
      case CTr:
	printf("CT");
	break;
      case PBr:
	printf("PB");
	break;
      case PTr:
	printf("PT");
	break;
      case SBr:
	printf("SB");
	break;
      case STr:
	printf("ST");
	break;
      case HBr:
	printf("HB");
	break;
      case HTr:
	printf("HT");
	break;
      case LBr:
	printf("LB");
	break;
      case L1r:
	printf("L1");
	break;
      case L2r:
	printf("L2");
	break;
      case L3r:
	printf("L3");
	break;
      case L4r:
	printf("L4");
	break;
      case L5r:
	printf("L5");
	break;
      case L6r:
	printf("L6");
	break;
      case CPr:
	printf("CP");
	break;
    }

    printf("%c",rightbracket);
  }


 void blankN() {
    printf("      ");
  }


void writeN (int n) {
    printf("(%d) ",n);
    if (n < 10)
      printf("  ");
    else if (n < 100)
      printf(" ");
  }

void writeD (int d) {
    printf("%d",d);
  }

void writePrimitive (int d) {
    switch (d) {
      case idDisplacement:
        printf("id      ");
	break;
      case notDisplacement:
	printf("not     ");
	break;
      case andDisplacement:
	printf("and     ");
	break;
      case orDisplacement:
	printf("or      ");
	break;
      case succDisplacement:
	printf("succ    ");
	break;
      case predDisplacement:
	printf("pred    ");
	break;
      case negDisplacement:
	printf("neg     ");
	break;
      case addDisplacement:
	printf("add     ");
	break;
      case subDisplacement:
	printf("sub     ");
	break;
      case multDisplacement:
	printf("mult    ");
	break;
      case divDisplacement:
	printf("div     ");
	break;
      case modDisplacement:
	printf("mod     ");
	break;
      case ltDisplacement:
	printf("lt      ");
	break;
      case leDisplacement:
	printf("le      ");
	break;
      case geDisplacement:
	printf("ge      ");
	break;
      case gtDisplacement:
	printf("gt      ");
	break;
      case eqDisplacement:
	printf("eq      ");
	break;
      case neDisplacement:
	printf("ne      ");
	break;
      case eolDisplacement:
	printf("eol     ");
	break;
      case eofDisplacement:
	printf("eof     ");
	break;
      case getDisplacement:
	printf("get     ");
	break;
      case putDisplacement:
	printf("put     ");
	break;
      case geteolDisplacement:
	printf("geteol  ");
	break;
      case puteolDisplacement:
	printf("puteol  ");
	break;
      case getintDisplacement:
	printf("getint  ");
	break;
      case putintDisplacement:
	printf("putint  ");
	break;
      case newDisplacement:
	printf("new     ");
	break;
      case disposeDisplacement:
	printf("dispose ");
	break;
    }
  }

 void writeInstruction (struct Data instr) {

    switch (getOperationCode(instr.operation)) {
      case LOADop:
	printf("LOAD  ");
	writeN(atoi(instr.length));
	writeD(atoi(instr.operand));
	writeR('[', getRegisterCode(instr.registers), ']');
	break;

      case LOADAop:
        printf("LOADA ");
        blankN();
        writeD(atoi(instr.operand));
        writeR('[', getRegisterCode(instr.registers), ']');
        break;

      case LOADIop:
        printf("LOADI ");
        writeN(atoi(instr.length));
        break;

      case LOADLop:
        printf("LOADL ");
        blankN();
        writeD(atoi(instr.operand));
        break;

      case STOREop:
        printf("STORE ");
        writeN(atoi(instr.length));
        writeD(atoi(instr.operand));
        writeR('[', getRegisterCode(instr.registers), ']');
        break;

      case STOREIop:
        printf("STOREI");
        writeN(atoi(instr.length));
        break;

      case CALLop:
        printf("CALL  ");
        if (getRegisterCode(instr.registers) == PBr) {
          blankN();
          writePrimitive(atoi(instr.operand));
        } else {
          writeR('(', atoi(instr.length), ')');
          printf("  ");
          writeD(atoi(instr.operand));
          writeR('[', getRegisterCode(instr.registers), ']');
        }
        break;

      case CALLIop:
        printf("CALLI ");
	break;

      case RETURNop:
        printf("RETURN");
        writeN(atoi(instr.length));
        writeD(atoi(instr.operand));
        break;

      case PUSHop:
        printf("PUSH  ");
        blankN();
        writeD(atoi(instr.operand));
        break;

      case POPop:
        printf("POP   ");
        writeN(atoi(instr.length));
        writeD(atoi(instr.operand));
        break;

      case JUMPop:
        printf("JUMP  ");
        blankN();
        writeD(atoi(instr.operand));
        writeR('[', getRegisterCode(instr.registers), ']');
        break;

      case JUMPIop:
        printf("JUMPI ");
        break;

      case JUMPIFop:
        printf("JUMPIF");
        writeN(atoi(instr.length));
        writeD(atoi(instr.operand));
        writeR('[', getRegisterCode(instr.registers), ']');
        break;

      case HALTop:
        printf("HALT  ");
    }

        printf("\n");
  }



void print_file(int cant){

    printf("The content of the file is:\n\n");

    for (int i = 0; i < cant; i++) {

        printf("%d: ",i);
        writeInstruction(object_array[i]);

    }
}


int load_Code(int cant){

    int pos ;
 

    for (pos = 0; pos < cant; pos++) {

        
        code[pos].op = getOperationCode(object_array[pos].operation);
        code[pos].r = getRegisterCode(object_array[pos].registers);
        code[pos].n = atoi(object_array[pos].length);
        code[pos].d = atoi(object_array[pos].operand);


    }

    return pos;

}

File: iteracion 3/js/loader.js



var json_parsed = null;


function getOperationCode( operation ){
   
    var opCode;

    switch (operation){

        case 'LOAD':
            opCode = machine_operation.LOADop;
            break;
        case "LOADA" :
            opCode = machine_operation.LOADAop;
            break;
        case "LOADI":
            opCode = machine_operation.LOADIop;
            break;
        case "LOADL":
            opCode = machine_operation.LOADLop;
            break;
        case "STORE":
            opCode = machine_operation.STOREop;
            break;
        case "STOREI":
            opCode = machine_operation.STOREIop;
            break;
        case "CALL":
            opCode = machine_operation.CALLop;
            break;
        case "CALLI":
            opCode = machine_operation.CALLIop;
            break;
        case "RETURN":
            opCode = machine_operation.RETURNop;
            break;
        case "PUSH":
            opCode = machine_operation.PUSHop;
            break;
        case "POP":
            opCode = machine_operation.POPop;
            break;
        case "JUMP":
            opCode = machine_operation.JUMPop;
            break;
        case "JUMPI":
            opCode = machine_operation.JUMPIop;
            break;
        case"JUMPIF":
            opCode = machine_operation.JUMPIFop;
            break;
        case"HALT":
            opCode = machine_operation.HALTop;
            break;
        default:
            opCode = -1;
                   
    }

    return opCode;

}
 
 
 function getRegisterCode( register ){
    
 
    var registerCode;

    switch (register){
    
         case "CB":
            registerCode = registers.CBr;
            break;
         case "CT":
            registerCode = registers.CTr;
            break;
         case "PB":
            registerCode = registers.PBr;
            break;
         case "PT":
            registerCode = registers.PTr;
            break;
         case "SB":
            registerCode = registers.SBr;
            break;
         case "ST":
            registerCode = registers.STr;
            break;
         case "HB":
            registerCode = registers.HBr;
            break;
         case "HT":
            registerCode = registers.HTr;
            break;
         case "LB":
            registerCode = registers.LBr;
            break;
         case "L1":
            registerCode = registers.L1r;
            break;
         case "L2":
            registerCode = registers.L2r;
            break;
         case "L3":
            registerCode = registers.L3r;
            break;
         case "L4":
            registerCode = registers.L4r;
            break;
         case "L5":
            registerCode = registers.L5r;
            break;
         case "L6":
            registerCode = registers.L6r;
            break;
         case "CP":
            registerCode = registers.CPr; 
            break;
         default: 
            registerCode = -1;
    
    }

    return registerCode;

 }

 function getRegisterName( r, ){
  
    var registerName;

    switch (r) {
      case registers.CBr:
        registerName = "CB";
        break;
      case registers.CTr:
        registerName ="CT";
	break;
      case registers.PBr:
        registerName ="PB";
	break;
      case registers.PTr:
        registerName ="PT";
	break;
      case registers.SBr:
        registerName ="SB";
	break;
      case registers.STr:
        registerName ="ST";
	break;
      case registers.HBr:
        registerName ="HB";
	break;
      case registers.HTr:
        registerName ="HT";
	break;
      case registers.LBr:
        registerName ="LB";
	break;
      case registers.L1r:
        registerName ="L1";
	break;
      case registers.L2r:
        registerName ="L2";
	break;
      case registers.L3r:
        registerName ="L3";
	break;
      case registers.L4r:
        registerName ="L4";
	break;
      case registers.L5r:
        registerName ="L5";
	break;
      case registers.L6r:
        registerName ="L6";
	break;
      case registers.CPr:
        registerName ="CP";
	break;
    }

    return registerName;
  
  } 




 function writeR ( leftbracket,  r,  rightbracket, data_string) {

    data_string += leftbracket;
    switch (r) {
      case "CB":
        data_string += "CB";
        break;
        case "CT":
        data_string += "CT";
	break;
      case "PB":
        data_string += "PB";
	break;
      case "PT":
        data_string += "PT";
	break;
      case "SB":
        data_string += "SB";
	break;
      case "ST":
        data_string += "ST";
	break;
      case "HB":
        data_string += "HB";
	break;
      case "HT":
        data_string += "HT";
	break;
      case "LB":
        data_string += "LB";
	break;
      case "L1":
        data_string += "L1";
	break;
      case "L2":
        data_string += "L2";
	break;
      case "L3":
        data_string += "L3";
	break;
      case "L4":
        data_string += "L4";
	break;
      case "L5":
        data_string += "L5";
	break;
      case "L6":
        data_string +="L6";
	break;
      case "CP":
        data_string +="CP";
	break;
    }

    data_string += rightbracket;
    return data_string;

  }



 function blankN(data_string) {
    data_string += "      ";
    return data_string;

}

function writeN (n,data_string) {
    data_string += "(" + n + ") ";
    if (n < 10)
        data_string +="  ";
    else if (n < 100)
        data_string += " ";

    return data_string;

  }

function writeD ( d,data_string) {
    data_string += d;
    return data_string;

  }

function writePrimitive (d,data_string) {
    switch (d) {
      case primitive_routines_address.idDisplacement:
        data_string +="id      ";
	break;
      case primitive_routines_address.notDisplacement:
        data_string +="not     ";
	break;
      case primitive_routines_address.andDisplacement:
        data_string +="and     ";
	break;
      case primitive_routines_address.orDisplacement:
        data_string +="or      ";
	break;
      case primitive_routines_address.succDisplacement:
        data_string +="succ    ";
	break;
      case primitive_routines_address.predDisplacement:
        data_string +="pred    ";
	break;
      case primitive_routines_address.negDisplacement:
        data_string +="neg     ";
	break;
      case primitive_routines_address.addDisplacement:
        data_string +="add     ";
	break;
      case primitive_routines_address.subDisplacement:
        data_string +="sub     ";
	break;
      case primitive_routines_address.multDisplacement:
        data_string +="mult    ";
	break;
      case primitive_routines_address.divDisplacement:
        data_string +="div     ";
	break;
      case primitive_routines_address.modDisplacement:
        data_string +="mod     ";
	break;
      case primitive_routines_address.ltDisplacement:
        data_string +="lt      ";
	break;
      case primitive_routines_address.leDisplacement:
        data_string +="le      ";
	break;
      case primitive_routines_address.geDisplacement:
        data_string +="ge      ";
	break;
      case primitive_routines_address.gtDisplacement:
        data_string +="gt      ";
	break;
      case primitive_routines_address.eqDisplacement:
        data_string += "eq      ";
	break;
      case primitive_routines_address.neDisplacement:
        data_string += "ne      ";
	break;
      case primitive_routines_address.eolDisplacement:
        data_string += "eol     ";
	break;
      case primitive_routines_address.eofDisplacement:
        data_string += "eof     ";
	break;
      case primitive_routines_address.getDisplacement:
        data_string += "get     ";
	break;
      case primitive_routines_address.putDisplacement:
        data_string += "put     ";
	break;
      case primitive_routines_address.geteolDisplacement:
        data_string += "geteol  ";
	break;
      case primitive_routines_address.puteolDisplacement:
        data_string += "puteol  ";
	break;
      case primitive_routines_address.getintDisplacement:
        data_string += "getint  ";
	break;
      case primitive_routines_address.putintDisplacement:
        data_string += "putint  ";
	break;
      case primitive_routines_address.newDisplacement:
        data_string += "new     ";
	break;
      case primitive_routines_address.disposeDisplacement:
        data_string += "dispose ";
	break;
    }

    return data_string;
  }




 function writeInstruction (instr) {

    var data_string = "";

    switch (instr.operation) {
      case "LOAD":
	data_string += "LOAD  ";
	data_string = writeN(instr.length,data_string);
	data_string =writeD(instr.operand,data_string);
	data_string =writeR('[', instr.register, ']',data_string);
	break;

      case "LOADA":
        data_string +="LOADA ";
        data_string = blankN(data_string);
        data_string = writeD(instr.operand,data_string);
        data_string = writeR('[', instr.register, ']',data_string);
        break;

      case "LOADI":
        data_string += "LOADI ";
        data_string = writeN(instr.length,data_string);
        break;

      case "LOADL":
        data_string += "LOADL ";
        data_string = blankN(data_string);
        data_string = writeD(instr.operand,data_string);
        break;

      case "STORE":
        data_string += "STORE ";
        data_string = writeN(instr.length,data_string);
        data_string = writeD(instr.operand,data_string);
        data_string = writeR('[', instr.register, ']',data_string);
        break;

      case "STOREI":
        data_string += "STOREI";
        data_string = writeN(instr.length,data_string);
        break;

      case "CALL":
        data_string += "CALL  ";
        if (getRegisterCode(instr.register) == registers.PBr) {
            data_string =  blankN(data_string);
            data_string = writePrimitive(instr.operand,data_string);
        } else {
            data_string = writeR('(', getRegisterName(instr.length), ')',data_string);
            data_string += "  ";
            data_string = writeD(instr.operand,data_string);
            data_string = writeR('[', instr.register, ']',data_string);
        }
        break;

      case "CALLI":
        data_string += "CALLI ";
	break;

      case "RETURN":
        data_string += "RETURN";
        data_string = writeN(instr.length,data_string);
        data_string = writeD(instr.operand,data_string);
        break;

      case "PUSH":
        data_string += "PUSH  ";
        data_string = blankN(data_string);
        data_string = writeD(instr.operand,data_string);
        break;

      case "POP":
        data_string += "POP   ";
        data_string = writeN(instr.length,data_string);
        data_string =  writeD(instr.operand,data_string);
        break;

      case "JUMP":
        data_string += "JUMP  ";
        data_string = blankN(data_string);
        data_string = writeD(instr.operand,data_string);
        data_string = writeR('[', instr.register, ']',data_string);
        break;

      case "JUMPI":
        data_string += "JUMPI ";
        break;

      case "JUMPIF":
        data_string += "JUMPIF";
        data_string = writeN(instr.length,data_string);
        data_string = writeD(instr.operand,data_string);
        data_string = writeR('[', instr.register, ']',data_string);
        break;

      case "HALT":
        data_string += "HALT  ";
    }

    displayContents(data_string);
  }



 function displayContents(contents) {
    var element = document.getElementById('file-content');
    element.textContent += contents;
  }
  
 function print_file(file_content){


    var size = file_content.length;

    element = document.getElementById('file_header');
    element.style.display = "inline-block"; 

    var element = document.getElementById('file-content');
    element.textContent = "";
    
    for ( var i = 0; i < size; i++) {

        var print_data = "  " + i.toString() + ": " 
        displayContents(print_data);
        writeInstruction(file_content[i]);
        displayContents("\n");

    }

}

document.getElementById('file-input').addEventListener('change', readFile, false)

 function readFile(e) {
    var file = e.target.files[0];
    if (!file) {
      return;
    }
    var reader = new FileReader();
    reader.onload = function(e) {
      var contents = e.target.result;
    json_parsed = JSON.parse(contents);
      print_file(json_parsed);
      
    };
    reader.readAsText(file);
  }


function load_Code(){

    var size = json_parsed.length;

    for (pos = 0; pos < size; pos++) {

        code[pos] = new Instruction(0,0,0,0);
        code[pos].op = getOperationCode(json_parsed[pos].operation);
        code[pos].r = getRegisterCode(json_parsed[pos].register);
        code[pos].n = parseInt(json_parsed[pos].length);
        code[pos].d = parseInt(json_parsed[pos].operand);

    }

    return pos;

}
